Disallow user group names starting with "@", "role:", "user:", "stream:".

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
- First commit is to add length restriction for user group name at API level.
- Second commit is to disallow user group names starting with "@", "role:", "user:", "stream:".
- Third commit is to remove "@" character from role-based system groups.

Fixes: <!-- Issue link, or clear description.--> #26148

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

sahil839 commented:
I am not sure why the production jobs are failing. Cannot find anything useful in the traceback.

I have not done any api documentation changes yet since there is an ongoing discussion [here](https://chat.zulip.org/#narrow/stream/412-api-documentation/topic/documenting.20error.20responses) and also we need to decide how do we want to handle the existing user groups using such prefixes.

PIG208 commented:
Looks like a networking error to `nodejs.org`.

PIG208 commented on zerver/lib/user_groups.py:
```diff
@@ -92,6 +92,13 @@ def access_user_group_for_setting(
 
+def check_user_group_name(group_name: str) -> str:
+    if len(group_name) > UserGroup.MAX_NAME_LENGTH:
+        raise JsonableError(_("Name too long!"))
```
It could be helpful to be more specific about the limit.
Like this example text from the developer documentation: https://github.com/zulip/zulip/blob/179d5cb37d2198a2a50e9903ffd74e436ca1820c/docs/tutorials/new-feature-tutorial.md?plain=1#L464

sahil839 replied:
I just took reference from the error message shown for user name. We can probably update that too if this change seems good for user groups.

timabbott replied:
The ideal version of this probably avoids including the number 100 in the string, instead using the variable, to save translation work if the number changes.

PIG208 commented on zerver/migrations/0458_remove_invalid_characters_from_system_group_name.py:
```diff
@@ -0,0 +1,34 @@
+    groups_to_update = []
+    for group in UserGroup.objects.filter(is_system_group=True):
+        old_group_name = group.name
+        group.name = old_group_name[1:]
```
Do we want to have a check here? In the rare case when this migration gets run twice, valid user group names get truncated as well.

sahil839 replied:
I added the check to be on the safer side.

PIG208 commented on zerver/lib/user_groups.py:
```diff
@@ -96,6 +96,10 @@ def check_user_group_name(group_name: str) -> str:
 
+    for invalid_prefix in UserGroup.INVALID_NAME_PREFIXES:
+        if group_name.startswith(invalid_prefix):
+            raise JsonableError(_("Invalid name for a user group."))
```
We could be more informative by returning the violating prefix.

timabbott commented on zerver/migrations/0458_remove_invalid_characters_from_system_group_name.py:
```diff
@@ -0,0 +1,38 @@
+        group.name = old_group_name[1:]
+        groups_to_update.append(group)
+
+    UserGroup.objects.bulk_update(groups_to_update, ["name"])
```
This migration I think should do a second pass where it renames any non-system groups that violate the naming rules established here. I checked Zulip Cloud and unsurprisingly nobody has named groups with names starting with things like `stream:`; group names starting with `@` are extremely rare and don't have a group in the same organization with `@` removed present.

So I'd propose that if a non-system-group has a name that violates the rules defined here, we should just rename it to remove the leading `@` (or more generally, whatever the problematic prefix is), and if that fails, rename it to `group_{id}`, where ID is its integer ID; I expect that fallback code will not run in practice but the safe fallback seems helpful to ensure the migration runs.

timabbott commented on zerver/models.py:
```diff
@@ -2241,6 +2241,7 @@ class PasswordTooWeakError(Exception):
 
 class UserGroup(models.Model):  # type: ignore[django-manager-missing] # django-stubs cannot resolve the custom CTEManager yet https://github.com/typeddjango/django-stubs/issues/1023
     MAX_NAME_LENGTH = 100
+    INVALID_NAME_PREFIXES = ["@", "role:", "user:", "stream:"]
```
Let's add `channel:` to this list for forward-safety if we rename `stream` to `channel` at some point.

timabbott commented:
This looks great; posted a handful of comments.

sahil839 commented on zerver/migrations/0458_remove_invalid_characters_from_user_group_name.py:
```diff
@@ -0,0 +1,77 @@
+                    )
+                    if len(matching_invalid_prefix) == 0:
+                        break
+                    group_name = group_name[len(matching_invalid_prefix) :]
```
This `while` loop is to make sure that we handle cases like `@role:admin` or `@@role:admin`.

PIG208 replied:
We can do `group_name = group_name.lstrip(matching_invalid_prefix)`.

sahil839 replied:
I tried using it initially but we cannot, `lstrip` strips all the characters and not a string. For example - if `group_name = stream:admin` and I call `group_name.lstrip("stream:")`, the result will be `dmin` because `stream:` contains `a` character and it is also removed.

PIG208 replied:
Ahh, I see. Thanks for explaining.

sahil839 commented on zerver/migrations/0458_remove_invalid_characters_from_user_group_name.py:
```diff
@@ -0,0 +1,77 @@
+                if group_name not in existing_group_names_set:
+                    group.name = group_name
+                else:
+                    group.name = f"group:{group.id}"
```
We are assuming here that there is not an existing group with such name.

timabbott replied:
Yeah, that seems probably fine.

sahil839 commented:
@timabbott updated the PR.
- Added code to disallow "channel:" prefix.
- Updated migration to handle non-system groups as well. Migration seems a bit complicated and hard to read as we need to handle the case like "@role:admin" and other cases where there are multiple invalid prefixes. Let me know if we can modify it in some way.

sahil839 commented on zerver/migrations/0458_remove_invalid_characters_from_user_group_name.py:
```diff
@@ -0,0 +1,77 @@
+    Realm = apps.get_model("zerver", "Realm")
+
+    groups_to_update = []
+    INVALID_NAME_PREFIXES = ["@", "role:", "user:", "stream:", "channel:"]
```
Yes, Updated the PR.

timabbott commented:
Fixed a mypy issue with the above; with those changes, this lgtm, marked to merge once CI passes. 

Thanks for cleaning this up @sahil839!
